File: docs/js/photoelectric.js
// Photoelectric effect — UI for py/photoelectric.py.

PROJECTS.push({
  id: 'photoelectric',
  name: '4. Photoelectric Effect',
  blurb: 'The stopping voltage V against light intensity for different metals. The solid line represents the real stopping voltages, whereas the dotted line is its extrapolation below the cutoff frequency. The four coloured lines represent the region of visible light.',

  async render(page) {
    const metals = await pyCall('photoelectric.web_metals');
    const lightFreqs = await pyCall('photoelectric.web_lights');
    // fixed across every metal, so switching material moves the curve, not the frame
    const [yMin, yMax] = await pyCall('photoelectric.web_yrange');
    const lightColours = ['#d62728', '#e6c700', '#2ca02c', '#1f77b4'];   // red yellow green blue

    let metal = metals[0];

    const plotDiv = el('div');
    const group = el('div', { class: 'radio-group' });
    for (const m of metals) {
      const inp = el('input', { type: 'radio', name: 'metal', value: m,
        onchange: () => { metal = m; draw(); } });
      if (m === metal) inp.checked = true;
      group.append(el('label', {}, inp, m));
    }

    page.append(el('div', { class: 'controls' }, group),
                el('div', { class: 'card' }, plotDiv));

    async function draw() {
      const d = await pyCall('photoelectric.web_curve', metal);

      const belowCutoff = d.V.map(v => v <= 0 ? v : null);
      const aboveCutoff = d.V.map(v => v > 0 ? v : null);

      // per-point extras for the hover box; full length so it lines up with the
      // null-masked y arrays
      const extra = d.frequencies.map((_, i) => [d.wavelength_nm[i], d.photon_eV[i]]);

      const common = 'f = %{x:.4~s}Hz<br>λ = %{customdata[0]} nm<br>' +
                     'photon energy = %{customdata[1]} eV<br>';

      Plotly.react(plotDiv, [
        { x: d.frequencies, y: belowCutoff, name: 'Extrapolated stopping voltage',
          mode: 'lines', line: { color: '#1f77b4', dash: 'dot' },
          customdata: extra,
          hovertemplate: common +
            'below cutoff: no emission<br>extrapolated V = %{y:.3f} V<extra></extra>' },
        { x: d.frequencies, y: aboveCutoff, name: 'Stopping voltage',
          mode: 'lines', line: { color: '#1f77b4' },
          customdata: extra,
          hovertemplate: common +
            'stopping voltage = %{y:.3f} V<br>max KE = %{y:.3f} eV<extra></extra>' },
        { x: [d.cutoff, d.cutoff], y: [yMin, yMax], name: `Cutoff frequency ${d.cutoff.toPrecision(3)} Hz`,
          mode: 'lines', line: { color: '#1f77b4', dash: 'dot', width: 1.5 },
          hovertemplate: `cutoff frequency = ${d.cutoff.toPrecision(4)} Hz<br>` +
            `threshold λ = ${d.cutoff_nm.toFixed(1)} nm<br>` +
            `W = ${d.W} eV<extra></extra>` }
      ], baseLayout({
        hovermode: 'closest',
        height: 460,
        title: { text: `Photoelectric effect for ${metal}:  W = ${d.W} eV` },
        xaxis: { title: { text: 'Frequency / Hz' }, range: [0, 2.5e15] },
        yaxis: { title: { text: 'Stopping voltage / V' }, range: [yMin, yMax], zeroline: true },
        shapes: lightFreqs.map((f, i) => ({
          type: 'line', x0: f, x1: f, y0: yMin, y1: yMax,
          line: { color: lightColours[i], width: 1.5 }
        }))
      }), PLOTCFG);
    }

    await draw();
  }
});

File: docs/py/photoelectric.py
# Physics from "Photoelectric Effect.py"; plotting is done by the JS layer.

import numpy as np

#Constants
h = 6.626e-34
e = 1.6021766e-19
c = 2.998e8

red = 4.4e14
yellow = 5.2e14
green = 5.8e14
blue = 6.6e14

metals = {
    "Silver (Ag)" : 4.3,
    "Aluminium (Al)" : 4.3,
    "Gold (Au)" : 5.1,
    "Copper (Cu)" : 4.7,
    "Tin (Sn)" : 4.4,
    "Lead (Pb)" : 4.3,
    "Tungsten (W)" : 4.5,
    "Nickel (Ni)" : 4.6,
    "Sodium (Na)" : 2.4
}


frequencies = np.linspace(0, 2.5e15, 500)


def stoppingVoltage(frequency, metal):
    global metals
    w = metals[metal] * e

    return (h*frequency)/e - (w/e)


# ---- adapter for the web UI ----

def web_metals():
    return list(metals.keys())


def web_lights():
    return [red, yellow, green, blue]


def web_yrange():
    # one fixed frame for every metal: only the curve should move, not the axes
    lo = min(stoppingVoltage(frequencies[0], m) for m in metals)
    hi = max(stoppingVoltage(frequencies[-1], m) for m in metals)
    return [lo - 0.5, hi + 0.5]


def web_curve(metal):
    stopVolts = stoppingVoltage(frequencies, metal)
    cutoff = metals[metal] * e/h
    return {
        "frequencies": frequencies.tolist(),
        "V": stopVolts.tolist(),
        "cutoff": cutoff,
        "W": metals[metal],
    }
